crates/client-api-messages/src/websocket.rs

```diff
@@ -184,6 +203,39 @@ pub struct Subscribe {
     pub request_id: u32,
 }
 
+/// Sent by client to register a subscription to single query, for which the client should receive
+/// receive relevant `TransactionUpdate`s.
+///
+/// After issuing a `SubscribeSingle` message, the client will receive a single
+/// `SubscribeApplied` message containing every current row which matches the query. Then, any
+/// time a reducer updates the query's results, the client will receive a `TransactionUpdate`
+/// containing the relevant updates.
+///
+/// If a client subscribes to queries with overlapping results, the client will receive
+/// multiple copies of rows that appear in multiple queries.
+#[derive(SpacetimeType)]
+#[sats(crate = spacetimedb_lib)]
+pub struct SubscribeSingle {
+    /// A single SQL `SELECT` query to subscribe to.
+    pub query: Box<str>,
+    /// An identifier for a client request.
+    /// This should not be reused for any other subscriptions on the same connection.
+    /// TODO: Can we call this subscription_id? It feels odd that this request_id will be used for multiple messages (the Unsubscribe call).
```

Comment:
I think there's some confusion here due to a typo in the proposal. The intended design is:

- Each request/response pair has a unique `request_id`, assigned by the client. It is not an error for the client to re-use `request_id`s, as their only purpose is to correlate a request/response pair; they have no semantic meaning to the host.
- This means that `Unsubscribe` gets a unique `request_id` distinct from the one in `SubscribeSingle`.
- The host looks at the `QueryId` included in the `Unsubscribe` message to determine which query is being removed.

It looks like at some point, someone (almost certainly me) made a copy-paste error in the proposal:

```rust
/// Server response to a client `Unsubscribe` request.
struct UnsubscribeApplied {
    /// Provided by the client via the `Subscribe` message.
    request_id: u32,
```

This should say, "Provided by the client via the `Unsubscribe` message." That is, the host treats it as an opaque nonce, and does not retain it or recognize it after sending its response.

Sorry for the confusion!